File: src/models/engram.py
"""
Engram Conditional Memory Architecture (DeepSeek-style O(1) Knowledge Bank).
Decouples static acoustic physics / material priors from dynamic neural computation.
"""

from typing import Tuple, Optional
import torch
import torch.nn as nn
import torch.nn.functional as F

DEFAULT_BANK_SIZE = 32768  # Option B: Rich configuration (32K entries, ~2MB)
DEFAULT_EMBED_DIM = 64
DEFAULT_HASH_HEADS = 4


class EngramBank(nn.Module):
    """
    O(1) hash-addressed static knowledge bank for physical audio priors.
    
    Attributes:
        bank_size: Number of stored discrete prior representations (32,768).
        embed_dim: Dimensionality of each stored embedding (64).
        num_hash_heads: Number of independent hash projections for collision resistance.
    """
    def __init__(
        self,
        bank_size: int = DEFAULT_BANK_SIZE,
        embed_dim: int = DEFAULT_EMBED_DIM,
        num_hash_heads: int = DEFAULT_HASH_HEADS,
    ):
        super().__init__()
        self.bank_size = bank_size
        self.embed_dim = embed_dim
        self.num_hash_heads = num_hash_heads

        # Static knowledge bank embedding table
        self.bank = nn.Embedding(bank_size, embed_dim)
        nn.init.normal_(self.bank.weight, mean=0.0, std=0.02)

        # Multi-head hash projections (fixed random orthogonal projections for deterministic hashing)
        hash_weights = []
        for i in range(num_hash_heads):
            gen = torch.Generator().manual_seed(42 + i * 1337)
            w = torch.randn(embed_dim, generator=gen)
            w = w / (torch.norm(w) + 1e-8)
            hash_weights.append(w)
        self.register_buffer("hash_projections", torch.stack(hash_weights, dim=0))  # (H, embed_dim)

        self.register_buffer(
            "hash_primes",
            torch.tensor([2654435761, 2246822519, 3266489917, 668265263], dtype=torch.int64)
        )

        # Gated fusion layer: combines retrieved engram knowledge with neural feature
        self.fuse_gate = nn.Linear(embed_dim * 2, embed_dim)
        self.out_norm = nn.LayerNorm(embed_dim)

    def compute_hash_indices(self, x: torch.Tensor) -> torch.Tensor:
        """
        Computes deterministic multi-head hash indices from continuous feature tensor.
        Input x: (*shape, embed_dim)
        Output: (*shape, num_hash_heads) of integer indices in [0, bank_size - 1].
        """
        # Linear projection per hash head: (*shape, num_hash_heads)
        projections = torch.matmul(x, self.hash_projections.t())

        # Quantize projection into integer bins
        scaled = (projections * 1000.0).to(torch.int64)

        primes = self.hash_primes[:self.num_hash_heads]
        indices = torch.abs((scaled * primes) % self.bank_size)
        return indices

    def lookup(self, indices: torch.Tensor) -> torch.Tensor:
        """
        Retrieves and aggregates embeddings across hash heads.
        indices: (*shape, num_hash_heads)
        Returns: (*shape, embed_dim)
        """
        retrieved = self.bank(indices)
        return torch.mean(retrieved, dim=-2)

    def forward(
        self,
        neural_features: torch.Tensor,
        query_key: Optional[torch.Tensor] = None,
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        """
        Performs O(1) engram retrieval and gated residual fusion.
        """
        query = neural_features if query_key is None else query_key
        indices = self.compute_hash_indices(query)
        engram_prior = self.lookup(indices)

        concat = torch.cat([neural_features, engram_prior], dim=-1)
        gate = torch.sigmoid(self.fuse_gate(concat))
        fused = gate * engram_prior + (1.0 - gate) * neural_features
        return self.out_norm(fused), gate

File: crates/inference/src/engram.rs
//! High-performance O(1) hash-addressed Engram knowledge bank for physical audio priors.
//!
//! Stores material acoustic physics, surface splash profiles, and HRTF priors.
//! Evaluated with allocation-free fixed-stack lookups.

pub const ENGRAM_BANK_SIZE: usize = 32768;
pub const ENGRAM_EMBED_DIM: usize = 64;
pub const ENGRAM_HASH_HEADS: usize = 4;

const HASH_PRIMES: [u64; 4] = [2654435761, 2246822519, 3266489917, 668265263];

/// Real-time Engram static knowledge bank.
#[derive(Clone, Debug)]
pub struct EngramBank {

    pub bank: Vec<[f32; ENGRAM_EMBED_DIM]>,
    pub hash_projections: [[f32; ENGRAM_EMBED_DIM]; ENGRAM_HASH_HEADS],
}

impl Default for EngramBank {
    fn default() -> Self {
        Self::new()
    }
}

impl EngramBank {
    pub fn new() -> Self {
        // Deterministic pseudo-random orthogonal hash heads matching Python model seed
        let mut hash_projections = [[0.0f32; ENGRAM_EMBED_DIM]; ENGRAM_HASH_HEADS];
        for (i, head) in hash_projections.iter_mut().enumerate() {
            let mut seed = (42 + i * 1337) as u64;
            let mut norm = 0.0f32;
            for val in head.iter_mut() {
                // Linear congruential generator for deterministic weights
                seed = seed.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
                let rand_f32 = ((seed >> 33) as f32) / (u32::MAX as f32) - 0.5;
                *val = rand_f32;
                norm += rand_f32 * rand_f32;
            }
            let inv_norm = 1.0 / norm.sqrt().max(1e-6);
            for val in head.iter_mut() {
                *val *= inv_norm;
            }
        }

        Self {
            bank: vec![[0.0f32; ENGRAM_EMBED_DIM]; ENGRAM_BANK_SIZE],
            hash_projections,
        }
    }

    /// Computes deterministic multi-head hash indices from query feature vector.
    pub fn compute_hash_indices(&self, query: &[f32; ENGRAM_EMBED_DIM]) -> [usize; ENGRAM_HASH_HEADS] {
        let mut indices = [0usize; ENGRAM_HASH_HEADS];
        for (h, idx) in indices.iter_mut().enumerate() {
            let proj: f32 = self.hash_projections[h]
                .iter()
                .zip(query.iter())
                .map(|(&w, &q)| w * q)
                .sum();
            let scaled = (proj * 1000.0).abs() as u64;
            *idx = ((scaled.wrapping_mul(HASH_PRIMES[h])) % (ENGRAM_BANK_SIZE as u64)) as usize;
        }
        indices
    }

    /// O(1) allocation-free retrieval across multi-head hash buckets.
    pub fn lookup(&self, query: &[f32; ENGRAM_EMBED_DIM], out: &mut [f32; ENGRAM_EMBED_DIM]) {
        let indices = self.compute_hash_indices(query);
        out.fill(0.0);
        let inv_heads = 1.0 / (ENGRAM_HASH_HEADS as f32);

        for &idx in &indices {
            let entry = &self.bank[idx.min(ENGRAM_BANK_SIZE - 1)];
            for (acc, &val) in out.iter_mut().zip(entry.iter()) {
                *acc += val * inv_heads;
            }
        }
    }

    /// Fused gated lookup: h_fused = gate * prior + (1 - gate) * feature.
    pub fn fuse_in_place(&self, features: &mut [f32; ENGRAM_EMBED_DIM], gate_alpha: f32) {
        let mut prior = [0.0f32; ENGRAM_EMBED_DIM];
        self.lookup(features, &mut prior);
        let alpha = gate_alpha.clamp(0.0, 1.0);
        for (f, &p) in features.iter_mut().zip(prior.iter()) {
            *f = alpha * p + (1.0 - alpha) * *f;
        }
    }
}
